// @playground-ai/cli/dist/http.js
export class CliHttpError extends Error {
    status;
    details;
    constructor(message, status, details) {
        super(message);
        this.name = "CliHttpError";
        this.status = status;
        this.details = details;
    }
}
function errorMessage(body, fallback) {
    if (!body || typeof body !== "object")
        return fallback;
    const b = body;
    if (typeof b.message === "string" && b.message.trim())
        return b.message;
    if (typeof b.code === "string" && typeof b.message === "string")
        return `${b.code}: ${b.message}`;
    if (typeof b.error === "string")
        return b.error;
    if (b.error && typeof b.error === "object") {
        const code = typeof b.error.code === "string" ? b.error.code : "ERROR";
        const message = typeof b.error.message === "string" ? b.error.message : fallback;
        return `${code}: ${message}`;
    }
    return fallback;
}
function authHeaders(auth) {
    const headers = {
        "Content-Type": "application/json",
    };
    if (auth.apiKey) {
        headers["X-API-Key"] = auth.apiKey;
        headers.Authorization = `Bearer ${auth.apiKey}`;
    }
    else if (auth.bearer) {
        headers.Authorization = `Bearer ${auth.bearer}`;
    }
    return headers;
}
export async function requestJson(input) {
    const response = await fetch(`${input.baseUrl}${input.path}`, {
        method: input.method ?? "GET",
        headers: authHeaders(input.auth),
        body: input.body !== undefined ? JSON.stringify(input.body) : undefined,
    });
    const text = await response.text();
    const contentType = response.headers.get("content-type") || "";
    const parsed = contentType.includes("application/json")
        ? (text ? JSON.parse(text) : {})
        : { message: text || response.statusText };
    if (!response.ok) {
        throw new CliHttpError(errorMessage(parsed, `Request failed (${response.status})`), response.status, parsed);
    }
    return parsed;
}
export async function requestSse(input) {
    const response = await fetch(`${input.baseUrl}${input.path}`, {
        method: "POST",
        headers: authHeaders(input.auth),
        body: JSON.stringify(input.body),
    });
    if (!response.ok) {
        const text = await response.text();
        let parsed = text;
        try {
            parsed = JSON.parse(text);
        }
        catch {
            // keep raw text fallback
        }
        throw new CliHttpError(errorMessage(parsed, `Streaming request failed (${response.status})`), response.status, parsed);
    }
    if (!response.body) {
        throw new CliHttpError("Stream ended before a body was returned", response.status);
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";
    while (true) {
        const { value, done } = await reader.read();
        if (done)
            break;
        buffer += decoder.decode(value, { stream: true });
        while (true) {
            const idx = buffer.indexOf("\n\n");
            if (idx < 0)
                break;
            const raw = buffer.slice(0, idx).trim();
            buffer = buffer.slice(idx + 2);
            if (!raw)
                continue;
            const lines = raw.split(/\r?\n/);
            let payload = "";
            for (const line of lines) {
                if (line.startsWith("data:"))
                    payload += line.slice(5).trim();
            }
            if (!payload)
                continue;
            if (payload === "[DONE]")
                return;
            try {
                await input.onEvent(JSON.parse(payload));
            }
            catch {
                await input.onEvent({ event: "raw", data: payload });
            }
        }
    }
}
